rfc-139-enable-http3.md

```diff
@@ -0,0 +1,135 @@
+
+### HTTP/3
+To solve the HOL blocking issue at the network layer, HTTP/3 uses a completely new transport layer protocol called QUIC. QUIC initially stood for "Quick UDP Internet Connections", but the IETF now don't use it as an acronym. It is simply the name of the protocol. QUIC is built upon User Datagram Protocol (UDP). Since QUIC uses UDP, it isn't bound by the restrictions that TCP imposes. So it has now been designed so that the network layer now understands the the concept of independent streams, which was restricted to the application layer under HTTP/2:
+
+![Difference between TCP and QUIC when downloading assets.](rfc-139/QUIC-illustration-final.jpg)
+
+In the illustration above, if packet 2 is lost (using HTTP/2), packets 3-8 are blocked until retransmission occurs. Using QUIC if packet 2 is lost only packets 1 and 3 are blocked since they exist in the same stream. The other streams can progress as normal since they are fully independent. QUIC doesn't require in-order packet transmission across the whole connection, but it does still retain ordering within each resource stream.
+
+## The effect of packet loss
+
+Using WebPageTest we can roughly simulate the effect that packet loss has on performance for our users. For testing I picking a fairly content heavy page like the [past Prime Ministers](https://www.gov.uk/government/history/past-prime-ministers) page and a low spec Moto G4 Android device. For the connection speed I've picked on one of the worse case scenarios from the [MLab data](https://docs.google.com/spreadsheets/d/1UTRucwLY2VC4944-cKunL9yUPJkjDKXDdIPk6GY3610/edit#gid=0). I'm emulating the connection the village of Colintraive has, which reportedly has a 450/230 Kbps connection speed. This is a connection speed equivalent to a 2G or slow 3G connection.
+
+### Visual filmstrip
+Below we can see a comparison of the page load under different packet loss conditions using HTTP/2. The top is the baseline, with 0% packet loss, progressing all the way down to 3% packet loss:
+
+![Comparing the visual filmstrip of the effect of packet loss on a page load](rfc-139/visual-filmstrip.png)
+
+As you can see, the visually complete metric sits at 9.5 seconds with no packet loss, but jumps to 13.5 seconds with 3% packet loss. This is a 42% increase.
+
+### Visual progress graph
+The visual progress graph clearly shows the effect packet loss has on page load times. As the packet loss increases, the page load slows down:
+
+![The visual progress graph comparing the percentage completeness of a page over time.](rfc-139/visual-progress.png)
+
+HTTP/3 + QUIC is designed to help improve the page load speed for users experiencing these connection conditions in the real world. 
```

Comment:
These charts are great for proving that a problem exists. But I was a bit surprised that there's no HTTP/3 equivalent for the charts - it would be great to see what effect enabling HTTP/3 would have on these page load times. If the answer is "eliminate packet loss", and therefore you'd expect a consistent 9.5 seconds under these conditions, then it would be worth clarifying that here.

I do also wonder if there may be a surprising change in page load time in the switch from TCP to UDP - we may not have packet loss, but could some pages still load more slowly than they would over HTTP/2?

Reply:
Yes I'd like to be able to conduct the same teat under HTTP/3 conditions. For this we'd need to ask Fastly to enable it on integration / staging first.

There is a question about how true the packet loss simulation is with WebPageTest (WPT), hence why it is a rough estimation. Packet loss tends to occur in bursts. So no PL for much of the connection, then a burst of 90% PL for a few seconds. I believe the way WPT models packet loss is that it takes a number of packets out of every 100, and loses them. e.g. lose 2 packets for every 100 sent for 2%. So the test above is great for showing the problem exists, but it's true representation and the fix HTTP/3 + QUIC could bring may be much harder to demonstrate. I think we will likely only see this in it's true form by looking at the web performance results for our "long-tail" users using real user monitoring (RUM) data.

But yes, it's a good question, and I'd like to see the results we get back.

Reply:
With RFC-147 implemented before HTTP/3 deployment on production, we should be able to compare H/2 vs H3 later in the year with SpeedCurve RUM data.